src/software/ai/navigator/path_planner/theta_star_path_planner.h

```diff
@@ -40,25 +41,120 @@ class ThetaStarPathPlanner : public PathPlanner
                                  const std::vector<ObstaclePtr> &obstacles) override;
 
    private:
-    class Coordinate : public std::pair<unsigned int, unsigned int>
+    class Coordinate
     {
        public:
         Coordinate(unsigned int row, unsigned int col)
-            : std::pair<unsigned int, unsigned int>(row, col)
+            : row_(row), col_(col), internal_comparison_key_(computeComparisonKey(*this))
         {
+            assert(row < (1 << 16));
+            assert(col < (1 << 16));
```

Comment:
Is this something that we might like to have throw an exception? We should probably at least add a comment explaining what these are checking.

Reply:
Alternatively, we could just have the constructor accept two `uint16_t`'s instead? Then this restriction is encoded in the types of the arguments.

Reply:
It is unlikely as long as `row` and `col` are positive since the maximum number of `row` and `col` are about `field_length/robot_radius`. The number shouldn't be in the order of `1<<16`. The assertion is included so that the `computeComparisonKey` would generate a unique key for each coordinate. The current logic is that the `row` value would occupy the low 16 bits of the key value and `col` would occupy the high 16 bits of the key value. When either one of the number exceed `1<<16` there could be unexpected behaviour. The original logic is to multiply `col` by the total number of row of the grid so that each coordinate can be mapped to a unique key. I think we considered similar approach as having the number of row and col of the grid as 2 argument inputs for constructor but it seems a bit weird that each coordinate know the number of row and col of the grid. I think there are other reasons but I forgot @jonathanlew

Reply:
I like the idea of accepting `uint16_t`s. I have a feeling these asserts are terrible for performance, so that's another benefit. If it was bad for performance then that could be causing the CI failures

Reply:
So if you compile with `-c opt` then the asserts are actually just compiled out entirely